File: kotlin/bc-envelope/src/main/kotlin/com/blockchaincommons/bcenvelope/EnvelopeEdge.kt
package com.blockchaincommons.bcenvelope

import com.blockchaincommons.bccomponents.Digest
import com.blockchaincommons.knownvalues.EDGE
import com.blockchaincommons.knownvalues.IS_A
import com.blockchaincommons.knownvalues.IS_A_RAW
import com.blockchaincommons.knownvalues.SOURCE
import com.blockchaincommons.knownvalues.SOURCE_RAW
import com.blockchaincommons.knownvalues.TARGET
import com.blockchaincommons.knownvalues.TARGET_RAW

/**
 * Edge support for Gordian Envelopes.
 *
 * Edges represent verifiable claims in an envelope graph structure
 * as defined in BCR-2026-003.
 */

/** Adds an edge envelope assertion. */
fun Envelope.addEdgeEnvelope(edge: Envelope): Envelope =
    addAssertion(EDGE, edge)

/** Returns all edge object envelopes. */
fun Envelope.edges(): List<Envelope> = objectsForPredicate(EDGE)

/** Validates an edge envelope's structure per BCR-2026-003. */
fun Envelope.validateEdge() {
    val inner = if (subject().isWrapped()) subject().unwrap() else this

    var seenIsA = false
    var seenSource = false
    var seenTarget = false

    for (assertion in inner.assertions()) {
        val predicate = try {
            assertion.tryPredicate().tryKnownValue().value
        } catch (_: Exception) {
            throw EnvelopeException.EdgeUnexpectedAssertion()
        }
        when (predicate) {
            IS_A_RAW -> {
                if (seenIsA) throw EnvelopeException.EdgeDuplicateIsA()
                seenIsA = true
            }
            SOURCE_RAW -> {
                if (seenSource) throw EnvelopeException.EdgeDuplicateSource()
                seenSource = true
            }
            TARGET_RAW -> {
                if (seenTarget) throw EnvelopeException.EdgeDuplicateTarget()
                seenTarget = true
            }
            else -> throw EnvelopeException.EdgeUnexpectedAssertion()
        }
    }

    if (!seenIsA) throw EnvelopeException.EdgeMissingIsA()
    if (!seenSource) throw EnvelopeException.EdgeMissingSource()
    if (!seenTarget) throw EnvelopeException.EdgeMissingTarget()
}

/** Extracts the 'isA' assertion object from an edge envelope. */
fun Envelope.edgeIsA(): Envelope {
    val inner = if (subject().isWrapped()) subject().unwrap() else this
    return inner.objectForPredicate(IS_A)
}

/** Extracts the 'source' assertion object from an edge envelope. */
fun Envelope.edgeSource(): Envelope {
    val inner = if (subject().isWrapped()) subject().unwrap() else this
    return inner.objectForPredicate(SOURCE)
}

/** Extracts the 'target' assertion object from an edge envelope. */
fun Envelope.edgeTarget(): Envelope {
    val inner = if (subject().isWrapped()) subject().unwrap() else this
    return inner.objectForPredicate(TARGET)
}

/** Extracts the edge's subject identifier. */
fun Envelope.edgeSubject(): Envelope {
    val inner = if (subject().isWrapped()) subject().unwrap() else this
    return inner.subject()
}

/** Filters edges by optional criteria. */
fun Envelope.edgesMatching(
    isA: Envelope? = null,
    source: Envelope? = null,
    target: Envelope? = null,
    subject: Envelope? = null,
): List<Envelope> {
    val allEdges = edges()
    return allEdges.filter { edge ->
        if (isA != null) {
            val edgeIsA = try { edge.edgeIsA() } catch (_: Exception) { return@filter false }
            if (!edgeIsA.isEquivalentTo(isA)) return@filter false
        }
        if (source != null) {
            val edgeSource = try { edge.edgeSource() } catch (_: Exception) { return@filter false }
            if (!edgeSource.isEquivalentTo(source)) return@filter false
        }
        if (target != null) {
            val edgeTarget = try { edge.edgeTarget() } catch (_: Exception) { return@filter false }
            if (!edgeTarget.isEquivalentTo(target)) return@filter false
        }
        if (subject != null) {
            val edgeSubject = try { edge.edgeSubject() } catch (_: Exception) { return@filter false }
            if (!edgeSubject.isEquivalentTo(subject)) return@filter false
        }
        true
    }
}

/**
 * A container for edge envelopes.
 */
class Edges {
    private val envelopes = mutableMapOf<Digest, Envelope>()

    /** Adds an edge envelope. */
    fun add(edgeEnvelope: Envelope) {
        envelopes[edgeEnvelope.digest()] = edgeEnvelope
    }

    /** Retrieves an edge by digest. */
    fun get(digest: Digest): Envelope? = envelopes[digest]

    /** Removes an edge by digest. */
    fun remove(digest: Digest): Envelope? = envelopes.remove(digest)

    /** Removes all edges. */
    fun clear() = envelopes.clear()

    /** Returns true if there are no edges. */
    fun isEmpty(): Boolean = envelopes.isEmpty()

    /** Returns the number of edges. */
    val size: Int get() = envelopes.size

    /** Returns an iterator over all edge envelopes. */
    val entries: Set<Map.Entry<Digest, Envelope>> get() = envelopes.entries

    /** Adds all edges as assertions to the given envelope. */
    fun addToEnvelope(envelope: Envelope): Envelope {
        var result = envelope
        for ((_, edgeEnvelope) in envelopes) {
            result = result.addAssertion(EDGE, edgeEnvelope)
        }
        return result
    }

    companion object {
        /** Extracts edges from an envelope. */
        fun fromEnvelope(envelope: Envelope): Edges {
            val edges = Edges()
            for (edgeEnv in envelope.edges()) {
                edges.envelopes[edgeEnv.digest()] = edgeEnv
            }
            return edges
        }
    }
}

/**
 * Interface for types that can have edges.
 */
interface Edgeable {
    fun edgesContainer(): Edges
    fun mutableEdgesContainer(): Edges
}

File: typescript/bc-envelope/src/envelope-edge.ts
import { Digest } from '@bc/components';
import {
    EDGE,
    IS_A,
    IS_A_RAW,
    SOURCE,
    SOURCE_RAW,
    TARGET,
    TARGET_RAW,
} from '@bc/known-values';

import { EnvelopeError } from './error.js';
import { Envelope } from './envelope.js';

export function addEdgeEnvelope(envelope: Envelope, edge: Envelope): Envelope {
    return envelope.addAssertion(EDGE, edge);
}

export function edges(envelope: Envelope): Envelope[] {
    return envelope.objectsForPredicate(EDGE);
}

export function validateEdge(envelope: Envelope): void {
    const inner = envelope.subject().isWrapped() ? envelope.subject().unwrap() : envelope;

    let seenIsA = false;
    let seenSource = false;
    let seenTarget = false;

    for (const assertion of inner.assertions()) {
        let predicateValue: bigint;
        try {
            predicateValue = assertion.tryPredicate().tryKnownValue().value;
        } catch {
            throw EnvelopeError.edgeUnexpectedAssertion();
        }

        if (predicateValue === IS_A_RAW) {
            if (seenIsA) {
                throw EnvelopeError.edgeDuplicateIsA();
            }
            seenIsA = true;
        } else if (predicateValue === SOURCE_RAW) {
            if (seenSource) {
                throw EnvelopeError.edgeDuplicateSource();
            }
            seenSource = true;
        } else if (predicateValue === TARGET_RAW) {
            if (seenTarget) {
                throw EnvelopeError.edgeDuplicateTarget();
            }
            seenTarget = true;
        } else {
            throw EnvelopeError.edgeUnexpectedAssertion();
        }
    }

    if (!seenIsA) throw EnvelopeError.edgeMissingIsA();
    if (!seenSource) throw EnvelopeError.edgeMissingSource();
    if (!seenTarget) throw EnvelopeError.edgeMissingTarget();
}

export function edgeIsA(envelope: Envelope): Envelope {
    const inner = envelope.subject().isWrapped() ? envelope.subject().unwrap() : envelope;
    return inner.objectForPredicate(IS_A);
}

export function edgeSource(envelope: Envelope): Envelope {
    const inner = envelope.subject().isWrapped() ? envelope.subject().unwrap() : envelope;
    return inner.objectForPredicate(SOURCE);
}

export function edgeTarget(envelope: Envelope): Envelope {
    const inner = envelope.subject().isWrapped() ? envelope.subject().unwrap() : envelope;
    return inner.objectForPredicate(TARGET);
}

export function edgeSubject(envelope: Envelope): Envelope {
    const inner = envelope.subject().isWrapped() ? envelope.subject().unwrap() : envelope;
    return inner.subject();
}

export function edgesMatching(
    envelope: Envelope,
    isA?: Envelope,
    source?: Envelope,
    target?: Envelope,
    subject?: Envelope,
): Envelope[] {
    return envelope.edges().filter((edge) => {
        if (isA != null) {
            try {
                if (!edge.edgeIsA().isEquivalentTo(isA)) {
                    return false;
                }
            } catch {
                return false;
            }
        }

        if (source != null) {
            try {
                if (!edge.edgeSource().isEquivalentTo(source)) {
                    return false;
                }
            } catch {
                return false;
            }
        }

        if (target != null) {
            try {
                if (!edge.edgeTarget().isEquivalentTo(target)) {
                    return false;
                }
            } catch {
                return false;
            }
        }

        if (subject != null) {
            try {
                if (!edge.edgeSubject().isEquivalentTo(subject)) {
                    return false;
                }
            } catch {
                return false;
            }
        }

        return true;
    });
}

export class Edges {
    readonly #envelopes = new Map<string, Envelope>();

    add(edgeEnvelope: Envelope): void {
        this.#envelopes.set(edgeEnvelope.digest().hex(), edgeEnvelope);
    }

    get(digest: Digest): Envelope | undefined {
        return this.#envelopes.get(digest.hex());
    }

    remove(digest: Digest): Envelope | undefined {
        const key = digest.hex();
        const existing = this.#envelopes.get(key);
        this.#envelopes.delete(key);
        return existing;
    }

    clear(): void {
        this.#envelopes.clear();
    }

    isEmpty(): boolean {
        return this.#envelopes.size === 0;
    }

    get size(): number {
        return this.#envelopes.size;
    }

    get entries(): ReadonlyArray<[string, Envelope]> {
        return [...this.#envelopes.entries()];
    }

    addToEnvelope(envelope: Envelope): Envelope {
        let result = envelope;
        for (const edgeEnvelope of this.#envelopes.values()) {
            result = result.addAssertion(EDGE, edgeEnvelope);
        }
        return result;
    }

    static fromEnvelope(envelope: Envelope): Edges {
        const edges = new Edges();
        for (const edgeEnvelope of envelope.edges()) {
            edges.#envelopes.set(edgeEnvelope.digest().hex(), edgeEnvelope);
        }
        return edges;
    }
}

export interface Edgeable {
    edgesContainer(): Edges;
    mutableEdgesContainer(): Edges;
}

declare module './envelope.js' {
    interface Envelope {
        addEdgeEnvelope(edge: Envelope): Envelope;
        edges(): Envelope[];
        validateEdge(): void;
        edgeIsA(): Envelope;
        edgeSource(): Envelope;
        edgeTarget(): Envelope;
        edgeSubject(): Envelope;
        edgesMatching(isA?: Envelope, source?: Envelope, target?: Envelope, subject?: Envelope): Envelope[];
    }
}

Envelope.prototype.addEdgeEnvelope = function addEdgeEnvelopeProto(this: Envelope, edge: Envelope): Envelope {
    return addEdgeEnvelope(this, edge);
};
Envelope.prototype.edges = function edgesProto(this: Envelope): Envelope[] {
    return edges(this);
};
Envelope.prototype.validateEdge = function validateEdgeProto(this: Envelope): void {
    validateEdge(this);
};
Envelope.prototype.edgeIsA = function edgeIsAProto(this: Envelope): Envelope {
    return edgeIsA(this);
};
Envelope.prototype.edgeSource = function edgeSourceProto(this: Envelope): Envelope {
    return edgeSource(this);
};
Envelope.prototype.edgeTarget = function edgeTargetProto(this: Envelope): Envelope {
    return edgeTarget(this);
};
Envelope.prototype.edgeSubject = function edgeSubjectProto(this: Envelope): Envelope {
    return edgeSubject(this);
};
Envelope.prototype.edgesMatching = function edgesMatchingProto(
    this: Envelope,
    isA?: Envelope,
    source?: Envelope,
    target?: Envelope,
    subject?: Envelope,
): Envelope[] {
    return edgesMatching(this, isA, source, target, subject);
};
